### PR TITLE
Tenth task add shopping cart model

### PR DESCRIPTION
**Pull request by apanyutin**

### CODE REVIEW

**Review by Elena-Bruyako — APPROVED** [review]
well done

File: src/main/java/com/example/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    @Operation(summary = "Get cart", description = "Get cart of user")
+    @GetMapping
+    public ShoppingCartDto getUserShoppingCart(Authentication authentication) {
+        return shoppingCartService.getShoppingCartByUserId(
+                getUserIdByAuthentication(authentication));
+    }
+
+    @Operation(summary = "Add book to cart", description = "Add book to the shopping cart")
+    @PostMapping
+    public CartItemDto addBookToShoppingCart(
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartDto addBookToShoppingCart(
```

File: src/main/java/com/example/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            @RequestBody @Valid CreateCartItemRequestDto requestDto,
+            Authentication authentication) {
+        requestDto.setShoppingCartId(getShoppingCartIdByAuthentication(authentication));
+        return cartItemService.save(requestDto);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/controller/ShoppingCartController.java]:
```suggestion
        return cartItemService.save(requestDto, getUserIdByAuthentication(authentication));
```

File: src/main/java/com/example/controller/ShoppingCartController.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    @Operation(summary = "Update book quantity", description = "Update book quantity in cart")
+    @PutMapping("/items/{cartItemId}")
+    public CartItemDto updateBookQuantityByCartItemId(
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartDto updateBookQuantityByCartItemId(Authentication authentication,
```

File: src/main/java/com/example/controller/ShoppingCartController.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            @PathVariable Long cartItemId,
+            @RequestBody @Valid UpdateCartItemRequestDto requestDto) {
+        return cartItemService.updateQuantityById(cartItemId, requestDto.getQuantity());
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/controller/ShoppingCartController.java]:
```suggestion
        return cartItemService.updateQuantityById(cartItemId, requestDto, getUserIdByAuthentication(authentication));
```

File: src/main/java/com/example/controller/ShoppingCartController.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    @Operation(summary = "Delete cartItem from cart", description = "Remove a book from cart")
+    @DeleteMapping("/items/{cartItemId}")
+    public void deleteByCartItemId(@PathVariable Long cartItemId) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/controller/ShoppingCartController.java:57]:
```suggestion
    public void deleteByCartItemId(@PathVariable Long cartItemId, Authentication authentication) {
```

File: src/main/java/com/example/controller/ShoppingCartController.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        cartItemService.deleteById(cartItemId);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/controller/ShoppingCartController.java]:
```suggestion
        cartItemService.deleteById(cartItemId, getUserIdByAuthentication(authentication));
```

File: src/main/java/com/example/controller/ShoppingCartController.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    private Long getUserIdByAuthentication(Authentication authentication) {
+        return ((User) authentication.getPrincipal()).getId();
+    }
+
+    private Long getShoppingCartIdByAuthentication(Authentication authentication) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/controller/ShoppingCartController.java]: remove

File: src/main/java/com/example/dto/cart/CreateCartItemRequestDto.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class CreateCartItemRequestDto {
+    private Long shoppingCartId;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/dto/cart/CreateCartItemRequestDto.java]:
```suggestion
```

File: src/main/java/com/example/dto/cart/CreateCartItemRequestDto.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+    @NotNull
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/dto/cart/CreateCartItemRequestDto.java:11]:
```suggestion
    @NotNull
    @Positive
```

File: src/main/java/com/example/dto/cart/CreateCartItemRequestDto.java [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    private Long bookId;
+    @Min(1)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/dto/cart/CreateCartItemRequestDto.java]:
```suggestion
    @Positive
```

File: src/main/java/com/example/dto/cart/CreateShoppingCartRequestDto.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package com.example.dto.cart;
+
+import jakarta.validation.constraints.NotNull;
+import java.util.Set;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class CreateShoppingCartRequestDto {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/dto/cart/CreateShoppingCartRequestDto.java]: remove

File: src/main/java/com/example/dto/cart/UpdateCartItemRequestDto.java [hunk 12]
```diff
@@ -0,0 +1,12 @@
+package com.example.dto.cart;
+
+import jakarta.validation.constraints.Min;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class UpdateCartItemRequestDto {
+    @Min(1)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/dto/cart/UpdateCartItemRequestDto.java]:
```suggestion
    @Positive
```

File: src/main/java/com/example/model/ShoppingCart.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
+
+    @OneToMany(mappedBy = "shoppingCart")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/model/ShoppingCart.java]:
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/com/example/model/CartItem.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+@Entity
+@Getter
+@Setter
+@Table(name = "carts_items")
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/com/example/model/CartItem.java [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+
+    @ManyToOne(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/com/example/repository/cart/ShoppingCartRepository.java [hunk 16]
```diff
@@ -0,0 +1,11 @@
+package com.example.repository.cart;
+
+import com.example.model.ShoppingCart;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.EntityGraph;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    @EntityGraph(attributePaths = "cartItems.book")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/repository/cart/ShoppingCartRepository.java]:
```suggestion
    @EntityGraph(attributePaths = {"cartItems", "cartItems.book"})
```

File: src/main/java/com/example/service/impl/CartItemServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+import com.example.model.CartItem;
+import com.example.repository.cart.CartItemRepository;
+import com.example.service.BookService;
+import com.example.service.CartItemService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class CartItemServiceImpl implements CartItemService {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/service/impl/CartItemServiceImpl.java]: replace this logic to ShopCartService

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/controller/ShoppingCartController.java (hunk 6 above)
Comment (apanyutin) [reply src/main/java/com/example/controller/ShoppingCartController.java]: Is it really need? Why just delete cart Item from table is not ok?

File: src/main/java/com/example/controller/ShoppingCartController.java [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+
+@Tag(name = "Shopping cart management", description = "Endpoints for managing cart")
+@RestController
+@RequestMapping("/cart")
+@RequiredArgsConstructor
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @Operation(summary = "Get cart", description = "Get cart of user")
+    @GetMapping
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/controller/ShoppingCartController.java:30]: Add `@PreAuthorize` where is needed

File: src/main/java/com/example/service/impl/ShoppingCartServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+    @Override
+    public ShoppingCartDto getCartByUserId(Long userId) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId).orElseThrow(
+                () -> new EntityNotFoundException(
+                        "Can't find shopping cart by user id = " + userId));
+        ShoppingCartDto shoppingCartDto = shoppingCartMapper.toDto(shoppingCart);
+        shoppingCartDto.setCartItems(shoppingCart.getCartItems().stream()
+                .map(cartItemMapper::toDto)
+                .collect(Collectors.toSet()));
+        return shoppingCartDto;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        return shoppingCartMapper.toDto(shoppingCartRepository.findByUserId(userId));
```

where in ShoppingCartRepository

```
@EntityGraph(attributePaths = {"cartItems", "cartItems.book"})
Optional<ShoppingCart> findByUserId(Long userId);
```

File: src/main/java/com/example/service/impl/ShoppingCartServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+        shoppingCart.getCartItems().add(cartItem);
+        shoppingCartRepository.flush();
+        return getCartByUserId(userId);
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto updateCartItemQuantityById(
+            Long cartItemId, UpdateCartItemRequestDto requestDto, Long userId) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId).orElseThrow(
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/service/impl/ShoppingCartServiceImpl.java:67]:
```suggestion
        ShoppingCart cart = shoppingCartRepository.findByUserId(userId);
        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(cartItemId, cart.getId())
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/controller/ShoppingCartController.java (hunk 18 above)
Comment (apanyutin) [reply src/main/java/com/example/controller/ShoppingCartController.java:30]: I don't realize why it need add @PreAuthorize User. This app don't have controllers methods without Admin access.

File: src/main/java/com/example/service/impl/ShoppingCartServiceImpl.java (hunk 20 above)
Comment (apanyutin) [reply src/main/java/com/example/service/impl/ShoppingCartServiceImpl.java:67]: I don't realize why it need to have cartId for find cartItem if it have separate table of cartItems with unique id (cartItemId)?

Comment (apanyutin) [reply src/main/java/com/example/service/impl/ShoppingCartServiceImpl.java:67]: https://i.imgur.com/4NzJyCD.png

**Pull request merged**